lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php

```diff
@@ -116,9 +116,23 @@ public function exec(string $statement) : int
     /**
      * {@inheritdoc}
      */
-    public function lastInsertId(?string $name = null) : string
+    public function lastInsertId() : string
     {
-        return db2_last_insert_id($this->conn);
+        $lastInsertId = db2_last_insert_id($this->conn);
+
+        if ($lastInsertId === '') {
+            throw new DB2Exception('No identity value was generated by the last statement.');
```

Comment:
Which is exactly why the original PR, #3368, was based on #3367, **which proposed to rework driver exceptions *before* implementing this separation**: a single `DriverException` class should be enough, the driver-specific classes serve no purpose.

I've now worked all the way back to make this PR compatible with the current state of exceptions, because of your [comment ](https://github.com/doctrine/dbal/pull/3367#issuecomment-443899736)back then:

> @Majkl578 drives the rework of exceptions. As far as I understand, design-wise it's not what he had in mind.

Reply:
This work has been recently merged as https://github.com/doctrine/dbal/pull/3712.

Reply:
Good! Now I would question why `DriverException` is an interface, and why we would need different implementations of it. Not all drivers have their own `DriverException` implementation, and there is, AFAICS, never a need to catch a specific driver exception.

Could you please read the intro of #3367 again, and tell me if you would like to proceed with this change first? I can make it compatible with the current `master` again, but it's quite a lot of work so I need to make sure that this is something you want, first.

Reply:
I'm not really up to speed on how exactly the exceptions are currently designed although I agree there are some flaws. E.g. the very fact of the existence of a class and an interface named `DriverException` makes me nervous.

> Good! Now I would question why `DriverException` is an interface, and why we would need different implementations of it.

This is true from one PoV: definitely, different implementations for MySQLi, PDO, etc. don't make much sense IMO. On the other hand, different implementations representing different domain-level seem to make sense to me. E.g. `class LastInsertIdNotSupported implements DriverException`.

> Could you please read the intro of #3367 again, and tell me if you would like to proceed with this change first? I can make it compatible with the current `master` again, but it's quite a lot of work so I need to make sure that this is something you want, first.

This is a circular dependency we've been trying to avoid :). For me, it would be easier if you rebased if first. Currently, I don't have enough time to be able to deeply familiarize myself with the approach based on the PR description (thanks for all the details) w/o being able to look at the diff.

Answering your question, I'd like to finish this patch first. If it makes more sense to leave these new exceptions as is, I'd rather leave them as is than switch context to the other ticket.